### PR TITLE
Input traj Update with navigation changes

### PR DESCRIPTION
**Pull request by SimoneMic**
Opening this PR for the integration of a custom method for computing the trajectory from an omnidirectional unicycle following a navigation path.
#
Points of discussion:
* A new `enum struct NavigationSetup` has been created for allowing multiple uses of the library. This is by default in `ManualMode` (i.e. usual joystick command in X and Y).
   This is configure by the method `setPlannerMode(...)`. This setup switches the call to the method of `computeNewSteps()` or `interpolateNewSteps()`
* Also a new `struct PoseStamped` has been created for tracking the poses computed in the trajectory.
* A new method `checkConstraints()` is being used in `interpolateNewSteps()` for checking the satisfaction of the constraints. It was taken by the tests code and improved.
* The navigation path is being passed by the method `setInputPath()` and stored in `std::vector<UnicycleState> m_inputPath`
* `m_currentController` has been moved to public (since needs to be accessed by the walking-controller by the TrajectoryGenerator) -> could create a dedicated method
* Still present some debug/info output that could be removed

We can discuss here what to improve and is yet to test this if it works with the actual `master` of the `walking-controller`

### CODE REVIEW

File: src/UnicycleGenerator.cpp [hunk 1]
```diff
@@ -386,9 +387,19 @@ bool UnicycleGenerator::generate(double initTime, double dT, double endTime)
     {
         std::lock_guard<std::mutex> guard(m_pimpl->mutex);
 
-        if (!(m_pimpl->planner->computeNewSteps(m_pimpl->leftFootPrint, m_pimpl->rightFootPrint, initTime, endTime))) {
-            std::cerr << "[UnicycleGenerator::generate] Failed to compute new steps." << std::endl;
-            return false;
+        if (m_pimpl->navigationConfig == NavigationSetup::NavigationMode && m_pimpl->planner->m_currentController == UnicycleController::DIRECT)
+        {
+            if (!(m_pimpl->planner->interpolateNewStepsFromPath(m_pimpl->leftFootPrint, m_pimpl->rightFootPrint, initTime, endTime))) {
+                std::cerr << "[UnicycleGenerator::generate] Failed to compute new steps." << std::endl;
```
Comment (mebbaid) [inline src/UnicycleGenerator.cpp]: Misleading print, perhaps a copy. May replace with `Failed to interpolate new steps from path`

File: src/UnicycleGenerator.cpp [hunk 2]
```diff
@@ -34,6 +34,7 @@ class UnicycleGenerator::UnicycleGeneratorImplementation {
     double mergePointRatioBegin = 0.5;
     double mergePointRatioEnd = 0.5;
 
+    NavigationSetup navigationConfig{NavigationSetup::ManualMode};
```
Comment (mebbaid) [inline src/UnicycleGenerator.cpp:37]: Are we fixing the Navigation mode? Are we not supposed to read this from a config file ?

File: tests/UnicycleTest.cpp [hunk 3]
```diff
@@ -153,7 +153,7 @@ bool plannerTest(){
     iDynTree::assertTrue(planner.setDesiredPersonDistance(conf.dX, conf.dY));
     iDynTree::assertTrue(planner.setPersonFollowingControllerGain(conf.K));
     iDynTree::assertTrue(planner.setMaximumIntegratorStepSize(conf.dT));
-    iDynTree::assertTrue(planner.setMaxStepLength(conf.maxL));
+    iDynTree::assertTrue(planner.setMaxStepLength(conf.maxL, 0.8));
```
Comment (mebbaid) [inline tests/UnicycleTest.cpp:156]: Can we parameterize 0.8?

**Review by mebbaid — CHANGES_REQUESTED** [review]
Thanks for the big work.
I left a few comments after skimming through the code that are mainly curiosities and trivial questions of mine.

File: tests/UnicycleTest.cpp (hunk 3 above)
Comment (SimoneMic) [reply tests/UnicycleTest.cpp:156]: I've copied this from the `main` branch of robotology
https://github.com/robotology/unicycle-footstep-planner/blob/f9db393b73ef92b2bf93ceac91da7a1db607f707/tests/UnicycleTest.cpp#L156

File: src/UnicycleGenerator.cpp (hunk 2 above)
Comment (SimoneMic) [reply src/UnicycleGenerator.cpp:37]: The `NavigationSetup` enum struct is set on the TrajectoryGenerator.cpp from the `walking-controllers` (red from a config file).
However, with this initialization we ensure retro-compatibility with the actual master branch from `walking-controller`

File: src/UnicycleGenerator.cpp (hunk 1 above)
Comment (SimoneMic) [reply src/UnicycleGenerator.cpp]: Fixed, thanks

File: src/UnicyclePlanner.cpp [hunk 4]
```diff
@@ -962,4 +1058,478 @@ bool UnicyclePlanner::setUnicycleController(UnicycleController controller)
...
+        double distance = std::sqrt(pow(navigationPath[i+1].position(0) - navigationPath[i].position(0), 2) +
+                                    pow(navigationPath[i+1].position(1) - navigationPath[i].position(1), 2) 
+                                    );
+        double slope = (navigationPath[i+1].position(1) - navigationPath[i].position(1)) / (navigationPath[i+1].position(0) - navigationPath[i].position(0));
+        double slope_angle = atan(slope);
+        double cos_theta = std::abs(cos(slope_angle));    
+        double sin_theta = std::abs(sin(slope_angle)); 
+
+        //if both poses aligned to the path = 0, if both poses aligned but not with the path > 0, if both poses perpendicular to the path = 1
+        double proprotionFactor = (std::abs(slope_angle - navigationPath[i+1].angle) + std::abs(slope_angle - navigationPath[i].angle))/ 2 / M_PI_2 ;
```
Comment (mebbaid) [inline src/UnicyclePlanner.cpp:1179]: according to [the MS documentation](https://learn.microsoft.com/en-us/cpp/c-runtime-library/math-constants?redirectedfrom=MSDN&view=msvc-170) ,  to use the `M_PI` , you need additional MS stuff, as in doing in the header also the line 
`_USE_MATH_DEFINES`
before doing `#include cmath`

**Comment by mebbaid** [comment]
@SimoneMic , looking at the logs of the CI tests, I found the following

```
2023-04-04T07:28:32.3224826Z D:\a\unicycle-footstep-planner\unicycle-footstep-planner\src\UnicyclePlanner.cpp(1179,142): error C2065: 'M_PI_2': undeclared identifier [D:\a\unicycle-footstep-planner\unicycle-footstep-planner\build\UnicyclePlanner.vcxproj]
2023-04-04T07:28:32.3228938Z D:\a\unicycle-footstep-planner\unicycle-footstep-planner\src\UnicyclePlanner.cpp(1190,31): error C2065: 'M_PI': undeclared identifier [D:\a\unicycle-footstep-planner\unicycle-footstep-planner\build\UnicyclePlanner.vcxproj]
2023-04-04T07:28:32.3232784Z D:\a\unicycle-footstep-planner\unicycle-footstep-planner\src\UnicyclePlanner.cpp(1194,60): error C2065: 'M_PI': undeclared identifier [D:\a\unicycle-footstep-planner\unicycle-footstep-planner\build\UnicyclePlanner.vcxproj]
2023-04-04T07:28:32.3236354Z D:\a\unicycle-footstep-planner\unicycle-footstep-planner\src\UnicyclePlanner.cpp(1198,60): error C2065: 'M_PI': undeclared identifier [D:\a\unicycle-footstep-planner\unicycle-footstep-planner\build\UnicyclePlanner.vcxproj]
2023-04-04T07:28:32.3239835Z D:\a\unicycle-footstep-planner\unicycle-footstep-planner\src\UnicyclePlanner.cpp(1210,32): error C2065: 'M_PI': undeclared identifier [D:\a\unicycle-footstep-planner\unicycle-footstep-planner\build\UnicyclePlanner.vcxproj]
2023-04-04T07:28:32.3247158Z D:\a\unicycle-footstep-planner\unicycle-footstep-planner\src\UnicyclePlanner.cpp(1212,61): error C2065: 'M_PI': undeclared identifier [D:\a\unicycle-footstep-planner\unicycle-footstep-planner\build\UnicyclePlanner.vcxproj]
2023-04-04T07:28:32.3259797Z D:\a\unicycle-footstep-planner\unicycle-footstep-planner\src\UnicyclePlanner.cpp(1224,35): warning C4244: 'initializing': conversion from 'double' to 'int', possible loss of data [D:\a\unicycle-footstep-planner\unicycle-footstep-planner\build\UnicyclePlanner.vcxproj]
2023-04-04T07:28:32.3261283Z D:\a\unicycle-footstep-planner\unicycle-footstep-planner\src\UnicyclePlanner.cpp(1235,39): error C2065: 'M_PI': undeclared identifier [D:\a\unicycle-footstep-planner\unicycle-footstep-planner\build\UnicyclePlanner.vcxproj]
2023-04-04T07:28:32.3263544Z D:\a\unicycle-footstep-planner\unicycle-footstep-planner\src\UnicyclePlanner.cpp(1237,42): error C2065: 'M_PI': undeclared identifier [D:\a\unicycle-footstep-planner\unicycle-footstep-planner\build\UnicyclePlanner.vcxproj]
2023-04-04T07:28:32.3265245Z D:\a\unicycle-footstep-planner\unicycle-footstep-planner\src\UnicyclePlanner.cpp(1239,45): error C2065: 'M_PI': undeclared identifier [D:\a\unicycle-footstep-planner\unicycle-footstep-planner\build\UnicyclePlanner.vcxproj]
2023-04-04T07:28:32.3267070Z D:\a\unicycle-footstep-planner\unicycle-footstep-planner\src\UnicyclePlanner.cpp(1241,42): error C2065: 'M_PI': undeclared identifier [D:\a\unicycle-footstep-planner\unicycle-footstep-planner\build\UnicyclePlanner.vcxproj]
2023-04-04T07:28:32.3268089Z D:\a\unicycle-footstep-planner\unicycle-footstep-planner\src\UnicyclePlanner.cpp(1334,27): error C2065: 'M_PI': undeclared identifier [D:\a\unicycle-footstep-planner\unicycle-footstep-planner\build\UnicyclePlanner.vcxproj]
2023-04-04T07:28:32.3273940Z D:\a\unicycle-footstep-planner\unicycle-footstep-planner\src\UnicyclePlanner.cpp(1336,51): error C2065: 'M_PI': undeclared identifier [D:\a\unicycle-footstep-planner\unicycle-footstep-planner\build\UnicyclePlanner.vcxproj]
2023-04-04T07:28:32.3788852Z   UnicycleBaseController.cpp
```

I dropped a comment of a possible fix.

**Comment by SimoneMic** [comment]
I added 
```
#define _USE_MATH_DEFINES
#include <cmath>
```
for the math constants in c++

**Review by mebbaid — APPROVED** [review]

**Comment by mebbaid** [comment]
> I added
> 
> ```
> #define _USE_MATH_DEFINES
> #include <cmath>
> ```
> 
> for the math constants in c++

Thanks @SimoneMic. Since also the PR was tested against `ergoCubGazeboV1`, I will merge.

**Pull request merged**